lib/Sema/CSSolver.cpp

```diff
@@ -110,7 +110,7 @@ Solution ConstraintSystem::finalize(
   Solution solution(*this, CurrentScore);
 
   // Update the best score we've seen so far.
-  if (solverState) {
+  if (solverState && !retainAllSolutions()) {
```

Comment:
I think so, because `BestScore` makes sense only if `worseThanBestSolution` check is allowed and I didn't want to make assertion check conditional.